be/src/exec/pipeline/scan_operator.h

```diff
@@ -58,12 +63,20 @@ class ScanOperator final : public SourceOperator {
     Status _trigger_next_scan(RuntimeState* state, int chunk_source_index);
     Status _try_to_trigger_next_scan(RuntimeState* state);
 
+    Status _capture_tablet_rowsets();
+
+private:
     // TODO(hcf) ugly, remove this later
     RuntimeState* _state = nullptr;
 
     bool _is_finished = false;
     int32_t _io_task_retry_cnt = 0;
 
+    // The row sets of tablets will become stale and be deleted, if compaction occurs
+    // and these row sets aren't referenced, which will typically happen when the tablets
+    // of the left table are compacted at building the right hash table. Therefore, reference
+    // the row sets into _tablet_rowsets in the preparation phase to avoid the row sets being deleted.
+    std::vector<std::vector<RowsetSharedPtr>> _tablet_rowsets;
```

Comment:
should put this code to olap scan operator.

Reply:
There isn't  class `OlapScanOperator`. `ScanOperator` just  corresponds to olap scan node.

Reply:
OK. we need to rename `ScanOperator` to `OlapScanOperator` in next PR